### PR TITLE
Implemented claw class and changed the end effector interface

### PR DESCRIPTION
**Pull request by denizlektemur**
The claw class has been implemented to work with the end effector interface.
The end effector interface has been changed slightly by moving the enum to internal comm and adding some functions so that the interface works with the can bus.

### CODE REVIEW

File: code/headers/claw.hpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    class claw_c : public end_effector_c {
+    public:
+        /**
+         * This is the constructor for the claw
+         * 
+         * @param pot_pin A hwlib adc pin to read the potmeter inside the servo of the claw default pin is a0
+         * 
+         * @param pwm_pin A r2d2 pwm lib pin used to send a pwm signal to the servo inside the claw default pin is 34
+         */
+        claw_c(hwlib::target::pin_adc pot_pin = hwlib::target::pin_adc(hwlib::target::ad_pins::a0),
```
Comment (LRstudentHU) [inline code/headers/claw.hpp]: I believe there is an interface available for the adc (e.g. hwlib::adc). Maybe you could receive that by reference? That way, you can write tests.

File: code/main.cpp [hunk 2]
```diff
@@ -11,10 +12,12 @@ int main(void) {
 
     r2d2::comm_c comm;
 
-    //no access to any end effector yet, using nullptr for the time being
-    auto end_effector = nullptr;
+    auto end_effector = r2d2::end_effectors::claw_c();
```
Comment (LRstudentHU) [inline code/main.cpp]: I don't understand why you create the pointer like this. On line 20, you could do `&end_effector` to pass it as a pointer.

File: code/src/claw.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    
+    int32_t claw_c::calc_pot_difference(uint8_t current_pwm) {
+        uint32_t pot_value;
+        uint32_t expected_pot;
+        int32_t difference;
+        pot_value = 0;
+        expected_pot = pot_per_pwm_step * (current_pwm-open_pwm) + pot_offset;
+        for(uint8_t i = 0; i < pot_scans; i++) {
+            pot_value += pot.read();
+            hwlib::wait_ms(1);
```
Comment (LRstudentHU) [inline code/src/claw.cpp]: Do you need to wait this long? Note that this will not do what you expect in an RTOS environment.

File: test/Makefile [hunk 4]
```diff
@@ -9,10 +9,10 @@
 #############################################################################
 
 # source files in this project (main.cpp is automatically assumed)
-SOURCES := $(wildcard ../code/src/*.cpp)
```
Comment (LRstudentHU) [inline test/Makefile:12]: Is this intentional?

**Review by LRstudentHU — CHANGES_REQUESTED** [review]

File: test/Makefile (hunk 4 above)
Comment (denizlektemur) [reply test/Makefile:12]: Yes if I use the wildcard lines travis gives an error about including the pwm library

File: code/src/claw.cpp (hunk 3 above)
Comment (denizlektemur) [reply code/src/claw.cpp]: The potmeter inside the servo needs some time to change its value. If I dont use a wait the reading will be inaccurate.

File: code/main.cpp (hunk 2 above)
Comment (denizlektemur) [reply code/main.cpp]: Good point I have changed it

File: code/src/claw.cpp (hunk 3 above)
Comment (LRstudentHU) [reply code/src/claw.cpp]: Yes, but do you need to wait for 1ms? Would 100msec give an inaccurate result? The module scheduler (HU-RTOS) will allow other tasks to run during this; if that shouldn't happen please use the busy wait

File: test/Makefile (hunk 4 above)
Comment (LRstudentHU) [reply test/Makefile:12]: Do you have a main file for tests?

**Review by LRstudentHU — CHANGES_REQUESTED** [review]

File: code/src/claw.cpp (hunk 3 above)
Comment (denizlektemur) [reply code/src/claw.cpp]: When testing it approximately needed 1ms to be consistent. I changed the wait to a busy wait.

File: test/Makefile (hunk 4 above)
Comment (denizlektemur) [reply test/Makefile:12]: I do have a main for tests but I havent changed anything in that main. It still has the standard code in it.

File: code/headers/claw.hpp (hunk 1 above)
Comment (denizlektemur) [reply code/headers/claw.hpp]: If I try it that way I get the error parameter of abstract class type hwlib::adc is not allowed.

Comment (LRstudentHU) [reply code/headers/claw.hpp]: `hwlib::adc` is an abstract class; that means you'll have to receive it by reference in the constructor. When instantiating your module, you'll pass the `hwlib::target::pin_adc` as normal.

**Review by LRstudentHU — APPROVED** [review]

**Pull request merged**